### DISCUSSION TITLE
Bevy Editor UI Pane System

### DISCUSSION THREAD

**Discussion by Atlas16A** [bevyengine/bevy_editor_prototypes#7] [Ideas]
Following the confirmation from Cart on direction for the UI layout system (Message from the discord, shown below) I have begun work on outlining rules, requirements, and implementation details while beginning to developed the system. This is a crosspost with a thread on the discord for the same purpose.

![SmartSelect_20240912_083005_Discord.jpg](https://github.com/user-attachments/assets/c690795c-6aaa-468f-a10a-9a08c95b4359)

Please tell me thoughts on blenders pane system and changes to it, either UX or UI related, remember this isn't about the contents of the panes, but the surrounding system.

# Blender's Pane model:
Summary based off Blender docs: https://docs.blender.org/manual/en/latest/interface/window_system/areas.html

- Areas in blender, also called Panes, are grouped/laid out in workspaces
- Where the mouse is hovering changes which pane is active and changes the behavior of certain shortcuts
- Resizing is done by the mouse grabbing from the edges of the Pane and dragging
- Pane Splitting/Merging are done by dragging the mouse from the corners of panes
- Merging panes will shade the pane that will be taken over
- The Cursor icon changes to help indicate what will happen when dragging
- The contents of the panes are selected from a drop down icon menu holding all the types of panes, with the current type's icon displayed on the drop down
- Areas can be swapped when pairing normal dragging with a shortcut

## Bevy Implementation: 

- Using Observers to know where the mouse is and the appropriate state for resizing/splitting/merging panes
- Panes act similar to a root UINode, with some extra requirements such as a header  to hold the pane type dropdown etc
- An api will exist for crates/plugins to define new panes without making custom implementations for most things.
- Styling of widgets will be owned by the editor and only custom widgets will be able to go outside of that, this ensures visual consistency in the style.
- At any given time a one pane is considered "active" this pane is whatever the mouse is hovering within, active panes scope the shortcuts to ensure shortcut overlap is possible without conflicts.
- All behaviours the mouse can do to the Pane system, a shortcut needs to be available to do that same behavior.
- Accessiblity needs to be considered, Need input on this for implementation approach.
- Pane layout is internal to each pane, there are few rules beyond nothing in a pane can exit that pane.

# Blender's Workspaces: 
Workspaces in blender are described as being preset layouts of the pane system and displayed at the top of the window.
https://docs.blender.org/manual/en/latest/interface/window_system/workspaces.html
## Bevy implementation:
- Stored as BSN
- Contains a list of only position/size/and pane type. 
- Loaded from a directory and automatically added to the top of the window

**Comment by janhohenheim** [comment]
I want to add that I really like how I can easily add and remove areas in Blender

**Reply by Atlas16A** [reply]
I do like how easy it is, however its not always clear where my mouse needs to be, so I'm thinking on adjusting the visuals to have more clear "put mouse here" spots.
